crates/fuel-core/src/query/balance.rs

```diff
@@ -23,15 +27,24 @@ use std::{
 
 pub mod asset_query;
 
-pub struct BalanceQueryContext<'a>(pub &'a Database);
+pub trait BalanceQueryData: Send + Sync {
+    fn balance(&self, owner: Address, asset_id: AssetId)
+        -> StorageResult<AddressBalance>;
 
-impl BalanceQueryContext<'_> {
-    pub fn balance(
+    fn balances(
+        &self,
+        owner: Address,
+        direction: IterDirection,
+    ) -> BoxedIter<StorageResult<AddressBalance>>;
+}
+
+impl BalanceQueryData for Box<dyn DatabasePort> {
```

Comment:
This is unfortunately blocked by
```
        let amount = AssetQuery::new(
            &owner,
            &AssetSpendTarget::new(asset_id, u64::MAX, u64::MAX),
            None,
            self,
        )
 ```
 
 Just a couple lines below, as `self` cannot be passed alone because it is explicitly not sized, and even when trying to pass it as a reference I kept running into multiple sizing issues, and a Box pointer was the most simple way around this, if you do have a way that works locally feel free to commit it directly but I tried to get this to work both now and when writing it, and the sizing constraint blocks it from working